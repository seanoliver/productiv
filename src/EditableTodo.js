import React, { useState } from 'react';
import Todo from './Todo';
import TodoForm from './TodoForm';

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * State
 * - editing: boolean in editing state or not
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {
  const [editing, setEditing] = useState(false);

  /** Toggle if this is being edited */
  function toggleEdit() {
    const isEditing = editing ? false : true;
    setEditing(isEditing);
  }

  /** Call remove fn passed to this. */
  function handleDelete(evt) {
    remove(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    update({ ...formData });
    toggleEdit();
  }

  return (
    <div className="EditableTodo">
      {editing === true &&
        <TodoForm
          handleSave={handleSave}
          initialFormData={todo}
        />}
      {editing === false &&
        <div className="mb-3">
          <div className="float-end text-sm-end">
            <button
              className="EditableTodo-toggle btn-link btn btn-sm"
              onClick={toggleEdit}>
              Edit
            </button>
            <button
              className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
              onClick={handleDelete}>
              Del
            </button>
          </div>
          <Todo
            id={todo.id}
            title={todo.title}
            description={todo.description}
            priority={todo.priority}
          />
        </div>}
    </div>
  );
}

export default EditableTodo;
