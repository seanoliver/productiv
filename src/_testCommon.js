const TEST_TODOS = [{
	id: 'im-an-id',
	title: 'Test title',
	description: 'Test description',
	priority: 1,
},
{
	id: 'im-also-an-id',
	title: 'Test title 2',
	description: 'Test description 2',
	priority: 1,
},
{
	id: 'me-too',
	title: 'Test title 3',
	description: 'Test description 3',
	priority: 1,
}];

export { TEST_TODOS }