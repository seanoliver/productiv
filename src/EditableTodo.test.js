import { render, fireEvent } from '@testing-library/react';
import EditableTodo from './EditableTodo.js';
import { TEST_TODOS } from './_testCommon.js';

const TEST_TODO = TEST_TODOS[0];

describe('EditableToDo', function () {
	it('renders without crashing', function () {
		render(
			<EditableTodo
				todo={TEST_TODO}
				update={jest.fn()}
				remove={jest.fn()}
			/>
		);
	});

	it('matches snapshot in non-editing state', function () {
		const { container, debug } = render(
			<EditableTodo
				todo={TEST_TODO}
				update={jest.fn()}
				remove={jest.fn()}
			/>
		);
		expect(container).toMatchSnapshot();
	});

  it('matches snapshot in editing state', function () {
		const { container, debug } = render(
			<EditableTodo
				todo={TEST_TODO}
				update={jest.fn()}
				remove={jest.fn()}
			/>
		);
    fireEvent.click(container.querySelector('.EditableTodo-toggle'));
		expect(container).toMatchSnapshot();
	});

	it('toggles between todo and form', function () {
    const { container, debug } = render(
			<EditableTodo
				todo={TEST_TODO}
				update={jest.fn()}
				remove={jest.fn()}
			/>
		);

    expect(container.querySelector('.Todo')).toBeInTheDocument();
    expect(container.querySelector('.TodoForm')).not.toBeInTheDocument();

    fireEvent.click(container.querySelector('.EditableTodo-toggle'));

    expect(container.querySelector('.Todo')).not.toBeInTheDocument();
    expect(container.querySelector('.NewTodoForm')).toBeInTheDocument();

	});
});