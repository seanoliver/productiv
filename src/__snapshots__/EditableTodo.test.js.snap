// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableToDo matches snapshot in editing state 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <form
      class="NewTodoForm"
    >
      <div
        class="mb-3"
      >
        <input
          aria-label="Title"
          class="form-control"
          id="newTodo-title"
          name="title"
          placeholder="Title"
          value="Test title"
        />
      </div>
      <div
        class="mb-3"
      >
        <textarea
          aria-label="Description"
          class="form-control"
          id="newTodo-description"
          name="description"
          placeholder="Description"
        >
          Test description
        </textarea>
      </div>
      <div
        class="mb-3 d-flex justify-content-between"
      >
        <div
          class="w-75 d-flex justify-content-between"
        >
          <label
            class="d-inline-flex"
            for="newTodo-priority"
          >
            Priority:  
          </label>
          <select
            class="form-control form-control-sm d-inline-flex"
            id="newTodo-priority"
            name="priority"
          >
            <option
              value="1"
            >
              Ultra-Über
            </option>
            <option
              value="2"
            >
              Über
            </option>
            <option
              value="3"
            >
              Meh
            </option>
          </select>
        </div>
        <button
          class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
        >
          Gø!
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`EditableToDo matches snapshot in non-editing state 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b
            class="Todo-title"
          >
            Test title
          </b>
          <small
            class="Todo-priority"
          >
            (priority: 
            1
            )
          </small>
        </div>
        <div>
          <small
            class="Todo-description"
          >
            Test description
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
